deployment/terraform/agent.go

```diff
@@ -44,41 +46,84 @@ func (t *Terraform) generateLoadtestAgentConfig(output *terraformOutput) loadtes
 	}
 }
 
-func (t *Terraform) initLoadtest(extAgent *ssh.ExtAgent, ip string) error {
-	sshc, err := extAgent.NewClient(ip)
-	if err != nil {
-		return err
+func (t *Terraform) configureAndRunAgents(extAgent *ssh.ExtAgent, output *terraformOutput) error {
+	var uploadBinary bool
+	var packagePath string
+	if strings.HasPrefix(t.config.LoadTestDownloadURL, filePrefix) {
+		packagePath = strings.TrimPrefix(t.config.LoadTestDownloadURL, filePrefix)
+		info, err := os.Stat(packagePath)
+		if err != nil {
+			return err
+		}
+		if !info.Mode().IsRegular() {
+			return fmt.Errorf("load-test package path %s has to be a regular file", packagePath)
+		}
+		uploadBinary = true
 	}
 
-	mlog.Info("Populating DB", mlog.String("ip", ip))
-	cmd := "cd mattermost-load-test-ng && export PATH=$PATH:/usr/local/go/bin && go run ./cmd/ltagent init"
-	if out, err := sshc.RunCommand(cmd); err != nil {
-		return fmt.Errorf("error running ssh command: %s, out: %s, error: %w", cmd, out, err)
+	for _, val := range output.Agents.Value {
+		sshc, err := extAgent.NewClient(val.PublicIP)
+		if err != nil {
+			return err
+		}
+		mlog.Info("configuring agent", mlog.String("ip", val.PublicIP))
+		if uploadBinary {
+			// upload the local file with a -local prefix to avoid collision
+			// with the previously downloaded file
+			dstFile := "/home/ubuntu/mattermost-load-test-ng-local.tar.gz"
+			mlog.Info(fmt.Sprintf("Uploading binary file %s", packagePath))
+			if out, err := sshc.UploadFile(packagePath, dstFile, false); err != nil {
+				return fmt.Errorf("error uploading file %q, output: %q: %w", packagePath, string(out), err)
+			}
+			if out, err := sshc.RunCommand("rm -rf mattermost-load-test-ng && " +
+				"tar xzf mattermost-load-test-ng-local.tar.gz && " +
+				"mv $(ls -d */ | grep mattermost-load-test-ng) mattermost-load-test-ng"); err != nil {
+				return fmt.Errorf("error running command, got output: %q: %w", string(out), err)
+			}
+		}
+
+		mlog.Info("Uploading agent service file")
+		rdr := strings.NewReader(strings.TrimSpace(agentServiceFile))
+		if out, err := sshc.Upload(rdr, "/lib/systemd/system/ltagent.service", true); err != nil {
+			return fmt.Errorf("error uploading file, output: %q: %w", string(out), err)
+		}
+
+		mlog.Info("Uploading coordinator service file")
+		rdr = strings.NewReader(strings.TrimSpace(coordinatorServiceFile))
```

Comment:
Should we upload the file in `initAndRunCoordinator` ?

Reply:
I thought it doesn't harm to be able to run coordinator in all of the agent instances 😄

Reply:
Sure, np.

Reply:
> I thought it doesn't harm to be able to run coordinator in all of the agent instances

Yes, as long as it runs only on one at a time, otherwise we have a problem.  
There's always the option to move the coordinator to the metrics instance which might be less pressured. We'll see if it makes sense once we start running load-tests regularly.